nbgrader_to_canvas/upload_grades.py

```diff
@@ -0,0 +1,379 @@
+
+                except MissingEntry:
+                    nb_student_and_score['posted_grade'] = ''
+
+
+                # student.id will give us student's username, ie shrakibullah. we will need to compare this to
+                # canvas's login_id instead of user_id              
+
+                # convert nbgrader username to canvas id (integer)
+                canvas_student_id = self.canvas_students[nb_student.id]
+                nb_grade_data[canvas_student_id] = nb_student_and_score
+                counter += 1
+                update_status(self._form_nb_assign_name, self._course_id, completion=35+30*counter/self._num_students)
+            return nb_grade_data
+
+    def _get_assignment(self):
+        assignment = self._course.get_assignment(self._form_canvas_assign_id)
+        if assignment is None:
+            raise Exception('Invalid form canvas_assign_id')
+        return assignment
+
+    def _get_max_score(self):
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            nb_assignment = gb.find_assignment(self._form_nb_assign_name)
+            return nb_assignment.max_score
+
+    def _get_num_students(self):
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            return len(gb.students)
+
+    def _create_assignment(self, max_score):
+        return self._course.create_assignment({'name':self._form_nb_assign_name, 'published':'true', 'assignment_group_id':self._group, 'points_possible':max_score})
+        
+    def _delete_comments(self):
+        '''clears old comments from submission to prevent cluttering of the comment section in case of re-submission'''
+        update_status(self._form_nb_assign_name, self._course_id, completion=5, status='Removing Old Feedback')
+        counter = 0
+        auth_header = {'Authorization': 'Bearer ' + self._flask_session['api_key']}
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            for student in gb.students:
+                try:
+                    submission = self.assignment_to_upload.get_submission(self.canvas_students[student.id],include=['submission_comments'])
+                    
+                    for comment in submission.submission_comments:
+                        if comment['comment'] == 'Transferred from datahub.':
+                            response = requests.delete(
+                                "{}api/v1/courses/{}/assignments/{}/submissions/{}/comments/{}".format(
+                                    settings.API_URL, self._course_id, self.assignment_to_upload.id, student.id, comment['id']
+                                ),
+                                headers = auth_header
+                                )
+                except Exception as ex:
+                    raise ex
+                    app.logger.error('Error deleting comments: {}'.format(ex))
```

Comment:
This line may not be reached sinced `raise` called first and will exit function. Consider moving `app.logger` before `raise`

Reply:
do this